### PR TITLE
core: fix several bugs

### PR DESCRIPTION
**Pull request by NashBC**
### Description

1. Fix the fork order check
2. Fix get vote attestation
3. Fix reorg logic of fast finality

### Rationale

The pervious logic is wrong, we should fix them.

### Example

NA

### Changes

Notable changes: 
* NA

### CODE REVIEW

File: params/config.go [hunk 1]
```diff
@@ -765,12 +765,12 @@ func (c *ChainConfig) CheckConfigForkOrder() error {
 		{name: "mirrorSyncBlock", block: c.MirrorSyncBlock},
 		{name: "brunoBlock", block: c.BrunoBlock},
 		{name: "eulerBlock", block: c.EulerBlock},
+		{name: "bonehBlock", block: c.BonehBlock},
+		{name: "lynnBlock", block: c.LynnBlock},
 		{name: "berlinBlock", block: c.BerlinBlock},
```
Comment (Pythonberg1997) [inline params/config.go:770]: The following four lines can be deleted. That's what we do in develop branch

Comment (NashBC) [reply params/config.go:770]: Cool, I think we can rebase develop branch then.

**Comment by buddh0** [comment]
LGTM

File: core/headerchain.go [hunk 2]
```diff
@@ -138,25 +138,27 @@ func (hc *HeaderChain) getFinalizedNumber(header *types.Header) uint64 {
...
 
 	finalized := p.GetFinalizedHeader(hc, header, header.Number.Uint64()-ancestor.Number.Uint64())
-	curFinalized := p.GetFinalizedHeader(hc, curHeader, curHeader.Number.Uint64()-ancestor.Number.Uint64())
-	if finalized == nil || curFinalized == nil {
-		return false
+	curFinalized := p.GetFinalizedHeader(hc, current, current.Number.Uint64()-ancestor.Number.Uint64())
+	if finalized == nil || curFinalized == nil || finalized.Number.Uint64() == curFinalized.Number.Uint64() {
+		return forker.ReorgNeeded(current, header)
+	} else if finalized.Number.Uint64() > curFinalized.Number.Uint64() {
+		return true, nil
```
Comment (buddh0) [inline core/headerchain.go:158]: if(curFinalized.Number.Uint64()> ancestor.Number.Uint64()){
    we can log that conflicting blocks have been finalized here.
}

File: core/headerchain.go [hunk 3]
```diff
@@ -138,25 +138,27 @@ func (hc *HeaderChain) getFinalizedNumber(header *types.Header) uint64 {
...
-		return false
+		return forker.ReorgNeeded(current, header)
 	}
 
 	finalized := p.GetFinalizedHeader(hc, header, header.Number.Uint64()-ancestor.Number.Uint64())
-	curFinalized := p.GetFinalizedHeader(hc, curHeader, curHeader.Number.Uint64()-ancestor.Number.Uint64())
-	if finalized == nil || curFinalized == nil {
-		return false
+	curFinalized := p.GetFinalizedHeader(hc, current, current.Number.Uint64()-ancestor.Number.Uint64())
+	if finalized == nil || curFinalized == nil || finalized.Number.Uint64() == curFinalized.Number.Uint64() {
```
Comment (buddh0) [inline core/headerchain.go:155]: we can log that conflicting blocks have been finalized here
if both finalized and curFinalized are bigger than ancestor

Comment (NashBC) [reply core/headerchain.go]: fixed

**Review by buddh0 — APPROVED** [review]

**Pull request merged**